[HUDI-920] Support Incremental query for MOR table

garyli1019 opened the pull request:
## What is the purpose of the pull request

Support incremental query for MOR table
https://issues.apache.org/jira/browse/HUDI-920

## Brief change log

  - Add MergeOnReadIncrementalRelation
  - Add test on TestMORDatasource

## Verify this pull request

This change added tests and can be verified as follows:


  - *Added unit tests in TestMORDatasource*
 

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

garyli1019 commented:
@vinothchandar This PR is functionally working. This is able to replace the existing `IncrementalRelation`, but for the safe side, we can mark this as an experimental feature. Once we gain enough confidence, we can completely replace the existing `IncrementalRelation`.

vinothchandar commented on hudi-spark/src/main/scala/org/apache/hudi/DataSourceOptions.scala:
```diff
@@ -49,6 +49,7 @@ object DataSourceReadOptions {
   val QUERY_TYPE_SNAPSHOT_OPT_VAL = "snapshot"
   val QUERY_TYPE_READ_OPTIMIZED_OPT_VAL = "read_optimized"
   val QUERY_TYPE_INCREMENTAL_OPT_VAL = "incremental"
+  val QUERY_TYPE_MOR_INCREMENTAL_OPT_VAL = "mor_incremental"
```
we need to guard this with a flag. query types are fundamental to design. I prefer not to overload them

garyli1019 replied:
This option was supposed to distinguish when the user wants to run an incremental query on the MOR table but on the parquet files only. This was not really necessary because users can just define the timestamp range to achieve the same goal. So we can just use the tableType. If COW, use the old way, if MOR, use the new relation. Don't need an extra option anymore imo.

vinothchandar commented on hudi-spark/src/main/scala/org/apache/hudi/HoodieMergeOnReadRDD.scala:
```diff
@@ -102,6 +104,44 @@ class HoodieMergeOnReadRDD(@transient sc: SparkContext,
 
+  private def logFileIterator(split: HoodieMergeOnReadFileSplit,
+                             config: Configuration): Iterator[InternalRow] =
+  new Iterator[InternalRow] {
```
this does not seem to be indented correctly?

garyli1019 replied:
nice catch :)

vinothchandar commented on hudi-spark/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:
```diff
@@ -0,0 +1,209 @@
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    log.debug(s"buildScan requiredColumns = ${requiredColumns.mkString(",")}")
+    log.debug(s"buildScan filters = ${filters.mkString(",")}")
+    // config to ensure the push down filter for parquet will be applied.
```
should these be left to the user on how sparkSession/sqlContext is configured instead?

garyli1019 replied:
This is the tricky part of this PR. We need to force a filter on `_hoodie_commit_time`, like what we have https://github.com/apache/hudi/blob/master/hudi-spark/src/main/scala/org/apache/hudi/IncrementalRelation.scala#L165.

Currently, I couldn't find an elegant way to force the filter regardless of what query the user is running. Spark optimization sometimes will skip scanning the file and produce incorrect results(e.g. count()). If we add a filter in RDD, then we will get involved in an extra ser/deser step. The elegant way might be getting into the Spark planing, but at this point I have no idea about how to do it. The easiest way is to ask the user to add a `.filter()` when loading the dataset, but we definitely don't wanna do that :) 

My approach here is to force Spark to always scan the Base file and always apply the filter:
- For the Base file, force the pushdown filter and avoid using the default ParquetFileFormat reader(baseFileIterator), which will not scan the file when the user uses `df.count()` and will produce incorrect results.
- For log file, no need to filter base on commit time.

So we need to make sure the pushdown filter was on in every run if we keep this approach, otherwise the result will be incorrect.

vinothchandar commented on hudi-spark/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    filteredFileGroup.map(f => {
+      val baseFile = f.getLatestDataFile
```
IIUC, your approach here is to find all the file groups impacted in commit range, then pull affected records from the latest file slice? the mergeOnReadRDD will handle the merging of such an file slice, with the commit filters applied. 

let me know @garyli1019 if I am understanding this correctly. it will help me review the code. 

in the meantime, can we add a separate flag/option to turn on `new.incremental.relation=true` or sth, to control this and push the changes

garyli1019 replied:
correct, that's what I was trying to do here.
sure, keep only one incremental query type makes sense to me.

vinothchandar replied:
@garyli1019 one issue we need to make sure we handle is when the file group is pending compaction. when that happens, the base file is not present in the latest slice, but stitched together from the previous slice. @bhasudha is handling this in the Hive PR. worth taking a look and ensuring

garyli1019 replied:
I think we can share the implementation of this part for Hive and Spark. I will move this part to a util class.

garyli1019 replied:
Just to make sure. This only possible for the base file right? The latest log file will always be on the latest file slice?

garyli1019 replied:
@vinothchandar getting back to this PR... Now I get a little confused, so when we pull from a pending compaction commit, why should we care about the base file that was not in the range? We just read the log file right? From a user perspective, I don't expect to read commit earlier than the commit I defined as a starting point.

garyli1019 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/HoodieInputFormatUtils.java:
```diff
@@ -443,4 +444,45 @@ private static HoodieBaseFile refreshFileStatus(Configuration conf, HoodieBaseFi
+   * @return HashMap<partitionPath, HashMap<fileName, FileStatus>>
+   * @throws IOException
+   */
+  public static HashMap<String, HashMap<String, FileStatus>> listStatusForAffectedPartitions(
```
Copied and changed this method from #1817 , will apply this change to that PR after merged.

garyli1019 commented:
Ready for review. cc: @vinothchandar @bhasudha

garyli1019 commented:
Found a bug https://issues.apache.org/jira/browse/HUDI-1434
Need to fix the bug before making this PR work.

vinothchandar commented:
@garyli1019 any hopes for this PR to be completed in the next day or so? :)

garyli1019 commented:
> @garyli1019 any hopes for this PR to be completed in the next day or so? :)

@vinothchandar The current version should work with the newer version commit file, but I haven't added the option to merge with the parquet file as we discussed in the meeting. Would you do a round of review?

vinothchandar commented:
Sounds good. Will do a review, morning time PST. 
Could you please check why CI is unhappy

vinothchandar commented:
```
[INFO] 
[ERROR] Failures: 
[ERROR]   TestMORDataSource.testCount:309 expected: <142> but was: <138>
```
seems related?

garyli1019 commented:
> ```
> [INFO] 
> [ERROR] Failures: 
> [ERROR]   TestMORDataSource.testCount:309 expected: <142> but was: <138>
> ```
> 
> seems related?

hmm, the test was passed in my local. The last time when we were working on the MOR data source, we saw the same thing happened that Travis and local are producing different results.

vinothchandar commented:
>hmm, the test was passed in my local. 

typically, if the test depends on some aspect of the data generated, then it can pass locally and fail in CI, because the random number seed used for data generation is different. Otherwise, it should pass. esp for checking counts etc. I have not seen it being flaky for other reasons.

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/HoodieInputFormatUtils.java:
```diff
@@ -470,4 +471,45 @@ private static HoodieBaseFile refreshFileStatus(Configuration conf, HoodieBaseFi
+   * @return HashMap<partitionPath, HashMap<fileName, FileStatus>>
+   * @throws IOException
+   */
+  public static HashMap<String, HashMap<String, FileStatus>> listStatusForAffectedPartitions(
```
can we just have this in `hudi-spark` for now. thats the only module that needs to call this.

garyli1019 replied:
This can be shared by other engines later. If we move this to spark we need to switch to scala code, then move it back later when supporting other engines. We could save some effort if we leave it here?

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/HoodieInputFormatUtils.java:
```diff
@@ -470,4 +471,45 @@ private static HoodieBaseFile refreshFileStatus(Configuration conf, HoodieBaseFi
+  public static HashMap<String, HashMap<String, FileStatus>> listStatusForAffectedPartitions(
+      Path basePath, List<HoodieInstant> commitsToCheck, HoodieTimeline timeline) throws IOException {
+    // Extract files touched by these commits.
+    // TODO This might need to be done in parallel like listStatus parallelism ?
```
Can we redo this such that it can use the metadata table for obtaining the listing? You can see how this is done in HoodieParquetInputFormat.

garyli1019 replied:
Are you referring to RFC-15 that not being landed yet? The current implementation of `HoodieParquetInputFormat` is listing all files of affected partitions and then do the filtering later.

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/HoodieInputFormatUtils.java:
```diff
@@ -470,4 +471,45 @@ private static HoodieBaseFile refreshFileStatus(Configuration conf, HoodieBaseFi
   }
 
+  /**
+   * List affected file status based on given commits.
```
a little bit more clearer doc? does this method obtain all the file status that were affected by the list of commits to check?

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/HoodieInputFormatUtils.java:
```diff
@@ -470,4 +471,45 @@ private static HoodieBaseFile refreshFileStatus(Configuration conf, HoodieBaseFi
+          Path fullPath = relativeFilePath != null ? FSUtils.getPartitionPath(basePath, relativeFilePath) : null;
+          if (fullPath != null) {
+            if (partitionToFileStatusesMap.get(entry.getKey()).containsKey(fullPath.getName())) {
+              // If filesystem support Append. Update the FileStatus of log file if being appended.
```
you can just pick the latest such file now. it will have the latest log size using `getFileSizeInBytes()`. No need to do the addition here.

garyli1019 replied:
fixed.

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:
```diff
@@ -0,0 +1,218 @@
+  * Relation, that implements the Hoodie incremental view for Merge On Read table.
+  *
+  */
+class MergeOnReadIncrementalRelation(val sqlContext: SQLContext,
```
can we subclass `IncrementalRelation` and reuse some of the code esp lines 54-66 etc? worth it?

garyli1019 replied:
`IncrementalRelation` extends `TablesScan` and this extends `PrunedFilteredScan` so I think to keep them separate is better.

vinothchandar replied:
ack

vinothchandar commented on hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:
```diff
@@ -0,0 +1,218 @@
+    sqlContext.sparkSession.sessionState.conf.setConfString("spark.sql.parquet.filterPushdown", "true")
+    sqlContext.sparkSession.sessionState.conf.setConfString("spark.sql.parquet.recordLevelFilter.enabled", "true")
+    sqlContext.sparkSession.sessionState.conf.setConfString("spark.sql.parquet.enableVectorizedReader", "false")
+    val pushDownFilter = {
```
this is the secret sauce to do the filtering at the record level, I guess

garyli1019 replied:
yes, if the filterPushdown was disabled then the record level filter will not work.

vinothchandar reviewed (approved):
I made a high level pass, looks like a good start to me. I would have liked us to skip the data blocks in the logs, if they don't fall in the commit range, rather than do the merge fully and filter at the higher level. The hudi log storage layout has enough intelligence/information for us to push this further down. But this is a good start, from which we can improve upon. 

Please fix the smaller changes and once CI is stable/non-flaky. we can land this.

garyli1019 commented:
Thanks, will update this PR before Friday night PST.

garyli1019 commented:
@vinothchandar addressed comments. You are right, the inconsistency was caused by the test data generator. The CI was happy after I changed from `generateUpdates` to `generateUniqueUpdates`.

vinothchandar commented:
Thanks @garyli1019 ! You are awesome. 
 Will take a pass, after CI is happy. and land,

vinothchandar commented:
@garyli1019 looks like the spark3 build broke

```
 mcist -DskipITs -Dscala-2.12 -Dspark3
```

leads to. 

```
[INFO] /Users/vs/Code/hudi/hudi-spark-datasource/hudi-spark/src/main/java:-1: info: compiling
[INFO] /Users/vs/Code/hudi/hudi-spark-datasource/hudi-spark/src/main/scala:-1: info: compiling
[INFO] Compiling 20 source files to /Users/vs/Code/hudi/hudi-spark-datasource/hudi-spark/target/classes at 1610229821451
[ERROR] /Users/vs/Code/hudi/hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:143: error: object SparkHadoopUtil in package deploy cannot be accessed in package org.apache.spark.deploy
[ERROR]     SparkHadoopUtil.get.addCredentials(jobConf)
[ERROR]     ^
[ERROR] one error found
```

FYI. I will try to fix this

garyli1019 commented:
#2425 will fix the build

vinothchandar commented:
Seems like this test is flaky. :( 

```
[INFO] 
[ERROR] Failures: 
[ERROR]   TestMORDataSource.testCount:310 expected: <152> but was: <2>
```

https://api.travis-ci.com/v3/job/470167079/log.txt

garyli1019 commented:
> Seems like this test is flaky. :( @garyli1019
> 
> ```
> [INFO] 
> [ERROR] Failures: 
> [ERROR]   TestMORDataSource.testCount:310 expected: <152> but was: <2>
> ```
> 
> https://api.travis-ci.com/v3/job/470167079/log.txt

@vinothchandar This was the compaction consistency question I mentioned in the sync meeting. There are actually 2 new insert records suppose to be here, but we triggered inline compaction that compacted 150 records. This test was 2 in my local back in August but now it's 152. So I was wondering if the compaction logic was changed somehow. @n3nash any thoughts?
